### PR TITLE
Rebuild table on resize

### LINKED ISSUE #32: when the web page resizes - get taller - we show blank rows for the table until a scroll event
When the web page resizes - get taller - we show blank rows for the table until a scroll event:

screen shot 2018-10-01 at 11 07 50 am

### PR DESCRIPTION
**Pull request by DanTup**
Fixes #32.

### CODE REVIEW

File: lib/tables.dart [hunk 1]
```diff
@@ -61,6 +61,7 @@ class Table<T> extends Object with SetStateMixin {
     _spacerAfterVisibleRows = new CoreElement('tr');
 
     element.onScroll.listen((_) => _scheduleRebuild());
+    window.onResize.listen((_) => _scheduleRebuild());
```
Comment (devoncarew) [inline lib/tables.dart]: I think we'll need to remove this when the widget is disposed.

**Comment by DanTup** [comment]
Ok, this got a fair bit more complicated. Seems like we had two possibilities for non-visible tables:

1. Tables that are removed from the DOM and discarded (this happens when popping them off the stack in Memory view)
2. Tables that are removed from the DOM but kept in a Map and re-inserted into the DOM later (this happens for most tables as you switch between pages - we restore the DOM when you come back)

So what I implemented was:

- A `dispose()` on table that cleans up the resize sub
- The resize sub skips the resize if the element is not in the DOM (but it wasn't disposed - eg. it's in the map and may be restored later)
- An event for when an item is added to the DOM which triggers a rebuild (so if you resize the window then return to a table, it rebuilds)

The watching for DOM insertion uses a MutationObserver (which watches only while someone is listening for events). I couldn't find a simpler way, but apparently this way is fast (I don't know if we can reduce the scope of it though - it may fire for unrelated DOM insertions).

I'm not totally happy with this - I put the DOM insertion code into `OnAddedToDomMixin` to keep it out of `CoreElement` however it's a bit weird because:

- We can't have constructors on Mixins so the init code is inside a getter
- To access the Element we had to add an "interface" `HasCoreElement` and add `implements HasCoreElement` to the mixin (which means adding `@override` in Table that uses it)

If you have any suggestions on how to tidy this up (or just alternative ways entirely), I'm more than happy to change it!

File: lib/ui/elements.dart [hunk 2]
```diff
@@ -306,3 +306,37 @@ class TrustedHtmlTreeSanitizer implements NodeTreeSanitizer {
   @override
   void sanitizeTree(Node node) {}
 }
+
+abstract class HasCoreElement {
```
[the remark this answers is not in the document]
Comment (DanTup) [reply lib/ui/elements.dart]: It's not - it exists purely to allow us to ensure anything implementing OnAddedToDomMixin has a `CoreElement element` so we can access it in the checks in the mutation observer code. I couldn't come up with a better way :(

File: lib/ui/elements.dart [hunk 3]
```diff
@@ -306,3 +306,37 @@ class TrustedHtmlTreeSanitizer implements NodeTreeSanitizer {
...
+  CoreElement get element;
+}
+
+abstract class OnAddedToDomMixin implements HasCoreElement {
```
[the remark this answers is not in the document]
Comment (DanTup) [reply lib/ui/elements.dart]: Done!

File: lib/tables.dart [hunk 4]
```diff
@@ -94,6 +98,31 @@ class Table<T> extends Object with SetStateMixin {
...
+    // Monitor Window resizes but don't rebuild if we're not in the DOM.
+    _windowResizeSubscription = window.onResize.listen((_) {
+      if (!isInDom) {
+        return;
+      }
+      _scheduleRebuild();
+    });
+  }
+
+  StreamSubscription<void> addRemoveDomSubscription;
```
[the remark this answers is not in the document]
Comment (DanTup) [reply lib/tables.dart]: Done! Also removed "Remove" from the name, since it was only doing added.

**Comment by DanTup** [comment]
I've addressed the comments and rebased on master. There were some minor conflicts due to constructor but otherwise seems to all work. LMK if you're still happy with it.

**Comment by devoncarew** [comment]
lgtm!

**Pull request merged**